src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -41,22 +41,10 @@ int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag
     MPIDI_XPMEM_global.local_rank = my_local_rank;
     MPIDI_XPMEM_global.node_group_ptr = NULL;
 
-    /* Allocate temporary shared buffer to exchange xpmem segids with all
-     * local processes. TODO: is there a better way to exchange at init ? */
-    mpi_errno = MPIDU_shm_seg_alloc(num_local * sizeof(xpmem_segid_t),
-                                    (void **) &xpmem_segids, MPL_MEM_SHM);
-    if (mpi_errno)
-        MPIR_ERR_POP(mpi_errno);
-    mpi_errno = MPIDU_shm_seg_commit(&shm_seg,
-                                     &shm_seg_barrier,
-                                     num_local, my_local_rank, local_rank_0, rank, MPL_MEM_SHM);
-    if (mpi_errno)
-        MPIR_ERR_POP(mpi_errno);
-
-    xpmem_segids[my_local_rank] = MPIDI_XPMEM_global.segid;
-    mpi_errno = MPIDU_shm_barrier(shm_seg_barrier, num_local);
-    if (mpi_errno)
-        MPIR_ERR_POP(mpi_errno);
+    MPIR_CHKLMEM_MALLOC(xpmem_segids, xpmem_segid_t *,
+                        sizeof(xpmem_segid_t) * num_local, mpi_errno, "xpmem segids", MPL_MEM_SHM);
+
+    MPIR_SHM_allgather(&MPIDI_XPMEM_global.segid, 1, MPI_LONG_LONG, xpmem_segids, 1, MPI_LONG_LONG);
 
```

Comment:
It might be better to use `MPI_BYTE` with `sizeof(xpmem_segid_t)` as do you not know the actual type of `xpmem_segid_t`.